File: examples/python/mnemonic_message.py
#!/usr/bin/env python


"""
 Copyright 2007-2019 United States Government as represented by the
 Administrator of The National Aeronautics and Space Administration.
 No copyright is claimed in the United States under Title 17, U.S. Code.
 All Rights Reserved.
"""



""" 
 @file mnemonic_message.py
 
 This file contains an example outlining how to create and publish a
 GMSEC-compliant Mnemonic Data Value message to the middleware bus using the
 Messaging Interface Standardization Toolkit (MIST) namespace
 ConnectionManager class function calls available for such operations.

 It is also recommended that you run a subscriber application
 (i.e. GMSEC_API/bin/gmsub) to see the LOG Messages which are published to the
 bus.
"""

import libgmsec_python
import sys

GMSEC_SPEC_VERSION   = "2016.00"
MVAL_MESSAGE_SUBJECT = "GMSEC.MISSION.SATELLITE.MSG.MVAL.MNEMONIC_MESSAGE"


def main():

    if(len(sys.argv) <= 1):
        usageMessage = "usage: " + sys.argv[0] + " mw-id=<middleware ID>" 
        print usageMessage
        return -1


    # Load the command-line input into a GMSEC Config object
    # A Config object is basically a key-value pair map which is used to
    # pass configuration options into objects such as Connections,
    # ConnectionManagers, Subscribe and Publish function calls, Messages,
    # etc.
    config = libgmsec_python.Config()

    for arg in sys.argv[1:]:
        value = arg.split('=')
        config.addValue(value[0], value[1])

    # If it was not specified in the command-line arguments, set LOGLEVEL
    # to 'INFO' and LOGFILE to 'stdout' to allow the program report output
    # on the terminal/command line
    initializeLogging(config)

    # Print the GMSEC API version number using the GMSEC Logging
    # interface
    # This is useful for determining which version of the API is
    # configured within the environment
    libgmsec_python.logInfo(libgmsec_python.ConnectionManager.getAPIVersion())

    try:
        # Create a ConnectionManager object
        connManager = libgmsec_python.ConnectionManager(config)

        # Open the connection to the middleware
        libgmsec_python.logInfo("Opening the connection to the middleware server")
        connManager.initialize()

        # Output middleware client library version
        libgmsec_python.logInfo(connManager.getLibraryVersion())

        # Create all of the GMSEC Message header Fields which will
        # be used by all GMSEC Messages
                
        # Note: Since these Fields contain variable values which are
        # based on the context in which they are used, they cannot be
        # automatically populated using MistMessage.
        definedFields = libgmsec_python.FieldList() 

        missionField = libgmsec_python.StringField("MISSION-ID", "MISSION")
        # Note: SAT-ID-PHYSICAL is an optional header Field, according
        # to the GMSEC ISD.
        satIdField = libgmsec_python.StringField("SAT-ID-PHYSICAL", "SPACECRAFT")
        facilityField = libgmsec_python.StringField("FACILITY", "GMSEC Lab")
        componentField = libgmsec_python.StringField("COMPONENT", "mnemonic_message")

        definedFields.append(missionField)
        definedFields.append(satIdField)
        definedFields.append(facilityField)
        definedFields.append(componentField)

        # Use setStandardFields to define a set of header fields for
        # all messages which are created or published on the
        # ConnectionManager using the following functions:
        # createLogMessage, publishLog, createHeartbeatMessage,
        # startHeartbeatService, createResourceMessage, 
        # publishResourceMessage, or startResourceMessageService
        connManager.setStandardFields(definedFields)

        # Populate the Mnemonic Sample(s)
        mSample = libgmsec_python.MnemonicSample("MS1", libgmsec_python.I32Field("MS1", 15))
        mSample.setEUValue(libgmsec_python.F32Field("My EU", 15.0))
        mSample.setFlags(1)
        mSample.setLimit(libgmsec_python.MnemonicSample.RED_HIGH)
        # Implicitly set limit enable/disable with setting of limit
        mSample.setQuality(True)
        mSample.setStalenessStatus(False)
        mSample.setTextValue("15")

        mnemonic_samples = libgmsec_python.MnemonicSampleList() 
        mnemonic_samples.append(mSample)

        # Add the Mnemonic values to a Mnemonic object
        mnemonic = libgmsec_python.Mnemonic("M1", mnemonic_samples)
        statusField = libgmsec_python.I16Field("status", 5)
        mnemonic.setStatus(statusField)
        mnemonic.setUnits("units")

        # Build up the Schema ID -- This is used to identify the
        # specific type of MVAL message to construct
                
        schemaId = GMSEC_SPEC_VERSION + ".GMSEC.MSG.MVAL"

        # Construct an MVAL Message and add the Mnemonic values to it
        mvalMessage = libgmsec_python.MnemonicMessage(MVAL_MESSAGE_SUBJECT, schemaId, connManager.getSpecification())
        mvalMessage.addMnemonic(mnemonic)

        # Add the header fields to the MVAL message
        connManager.addStandardFields(mvalMessage)

        libgmsec_python.logInfo("Publishing MVAL message:\n" + mvalMessage.toXML())
        connManager.publish(mvalMessage)
        
    except libgmsec_python.Exception as e:
        libgmsec_python.logError(e.what())
        return -1

    return 0



def initializeLogging(config):
    
    logLevel  = config.getValue("LOGLEVEL")
    logFile   = config.getValue("LOGFILE")

    if (not logLevel):
       
        config.addValue("LOGLEVEL", "INFO")
     
    if (not logFile):
        
        config.addValue("LOGFILE", "STDERR")

#
# Main entry point of script
#
if __name__=="__main__":
    sys.exit(main())
        

File: examples/csharp-dotnet/mnemonic_message/mnemonic_message.cs
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/**
 * @file mnemonic_message.cs
 *
 * This file contains an example outlining how to create and publish a
 * GMSEC-compliant Mnemonic Data Value message to the middleware bus using the
 * Messaging Interface Standardization Toolkit (MIST) namespace
 * ConnectionManager class function calls available for such operations.
 *
 * It is also recommended that you run a subscriber application
 * (i.e. GMSEC_API/bin/gmsub) to see the LOG Messages which are published to the
 * bus.
 */

using GMSEC.API;
using GMSEC.API.UTIL;
using GMSEC.API.MIST;
using GMSEC.API.MIST.MESSAGE;
using System;
using System.Collections.Generic;

class mnemonic_message
{
	static private String GMSEC_SPEC_VERSION   = "201600";
	static private String MVAL_MESSAGE_SUBJECT = "GMSEC.MISSION.SATELLITE.MSG.MVAL.MNEMONIC_MESSAGE";
	static int Main(string[] args)
	{
		if (args.Length < 1)
		{
			Console.WriteLine("usage mnemonic_message.exe mw-id=<middleware ID>");
			return -1;
		}

		Config config = new Config(args);

		InitializeLogging(config);

		//o Set the GMSEC message specification version to be used to determine
		// what the structure of messages is for verification and the
		// construction of MistMessages
		config.AddValue("GMSEC-SPECIFICATION-VERSION", GMSEC_SPEC_VERSION);

		//o Enable Message validation.  This parameter is "false" by default.
		config.AddValue("GMSEC-MSG-CONTENT-VALIDATE", "true");

		Log.Info(ConnectionManager.GetAPIVersion());

		try
		{
			ConnectionManager connManager = new ConnectionManager(config);

			Log.Info("Opening the connection to the middleware server");
			connManager.Initialize();

			Log.Info(connManager.GetLibraryVersion());

			//o Create all of the GMSEC Message header Fields which will
			// be used by all GMSEC Messages
			//
			// Note: Since these Fields contain variable values which are
			// based on the context in which they are used, they cannot be
			// automatically populated using MistMessage.
			List<Field> definedFields = new List<Field>();

			StringField missionField = new StringField("MISSION-ID", "MISSION");
			// Note: SAT-ID-PHYSICAL is an optional header Field, according
			// to the GMSEC ISD.
			StringField satIdField = new StringField("SAT-ID-PHYSICAL", "SPACECRAFT");
			StringField facilityField = new StringField("FACILITY", "GMSEC Lab");
			StringField componentField = new StringField("COMPONENT", "device_message");

			definedFields.Add(missionField);
			definedFields.Add(satIdField);
			definedFields.Add(facilityField);
			definedFields.Add(componentField);

			//o Use setStandardFields to define a set of header fields for
			// all messages which are created or published on the
			// ConnectionManager using the following functions:
			// createLogMessage, publishLog, createHeartbeatMessage,
			// startHeartbeatService, createResourceMessage,
			// publishResourceMessage, or startResourceMessageService
			connManager.SetStandardFields(definedFields);

			//o Populate the Mnemonic Sample(s)
			MnemonicSample mSample = new MnemonicSample("MS1", new I32Field("MS1", 15));
			mSample.SetEUValue(new F32Field("My EU", (float) 15.0));
			mSample.SetFlags(1);
			mSample.SetLimit(MnemonicSample.LimitFlag.RED_HIGH);
			// Implicitly set limit enable/disable with setting of limit
			mSample.SetQuality(true);
			mSample.SetStalenessStatus(false);
			mSample.SetTextValue("15");

			List<MnemonicSample> mnemonic_samples = new List<MnemonicSample>();
			mnemonic_samples.Add(mSample);

			//o Add the Mnemonic values to a Mnemonic object
			Mnemonic mnemonic = new Mnemonic("M1", mnemonic_samples);
			I16Field statusField = new I16Field("status", 5);
			mnemonic.SetStatus(statusField);
			mnemonic.SetUnits("units");

			//o Determine which version of the GMSEC message specification
			// the ConnectionManager was initialized with
			uint version = connManager.GetSpecification().GetVersion();
			String gmsecSpecVersion = "";
			if (version == 201600)
			{
				gmsecSpecVersion = "2016.00";
			}
			else if (version == 201400)
			{
				gmsecSpecVersion = "2014.00";
			}

			//o Build up the Schema ID -- This is used to identify the
			// specific type of MVAL message to construct
			String schemaId = gmsecSpecVersion + ".GMSEC.MSG.MVAL";

			//o Construct an MVAL Message and add the Mnemonic values to it
			using(GMSEC.API.MIST.MESSAGE.MnemonicMessage mvalMessage = new GMSEC.API.MIST.MESSAGE.MnemonicMessage(MVAL_MESSAGE_SUBJECT, schemaId, connManager.GetSpecification()))
			{
				mvalMessage.AddMnemonic(mnemonic);

				//o If validating with the 2014 spec, the MSG-ID field is
				// required
				if (version == 201400)
				{
					mvalMessage.SetValue("MSG-ID", "MVAL Request MSG-ID would go here");
				}

				//o Add the header fields to the MVAL message
				connManager.AddStandardFields(mvalMessage);

				Log.Info("Publishing MVAL message:\n" + mvalMessage.ToXML());
				connManager.Publish(mvalMessage);
			}

			connManager.Cleanup();
		}
		catch (GMSEC_Exception e)
		{
			Log.Error(e.ToString());
			return -1;
		}

		return 0;
	}

	static void InitializeLogging(Config config)
	{
		String logLevel = config.GetValue("LOGLEVEL");
		String logFile = config.GetValue("LOGFILE");

		if (logLevel == null)
		{
			config.AddValue("LOGLEVEL", "INFO");
		}
		if (logFile == null)
		{
			config.AddValue("LOGFILE", "STDERR");
		}
	}
}
